hook_civicrm_links - Add docblocks and event-checker

totten opened the pull request:
Overview
----------------------------------------

`hook_civicrm_links` is fired in numerous ways (incl ~9 direct calls to `CRM_Utils_hook::links()`, ~96 indirect calls via `CRM_Core_Action::formLink()`). This adds some more docblocks and validation.

(This is an off-shoot of the discussion with @eileenmcnaughton  on https://github.com/civicrm/civicrm-core/pull/26243/files#r1328057493. It seems to me better to have the tests complaining about quirks in `hook_links` rather than random web-pages.)

Before
----------------------------------------

* Docblock on `links()` doesn't explain what's expected
* No systemic validation mechanism

After
----------------------------------------

* Docblock on `links()` (derived from https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_links/)
* Anytime the hook fires within a unit-test, it will validate the data passing through `$links`

Comments
----------------------------------------

* Someday, I'd like to have a flag or extension that we use on dev-sites -- where it runs the event-checker during regular web-browsing...
* I did a partial test-run locally, and it shows some more properties that weren't previously documented. Gonna let CI do a full run and see how many other bits it finds.

totten committed:
hook_civicrm_links - Add docblocks and event-checker

eileenmcnaughton commented:
Cool tests are good

eileenmcnaughton commented:
you should probably merge this though - https://github.com/civicrm/civicrm-core/pull/27480 to reduce the false positives

colemanw commented:
retest this please

totten committed:
pcp.user.actions - Decompose deviant list of links

The problem is that:

* `hook_links('pcp.user.actions')` fires with a malformed set links
* The malformed set of links has been specifically encoded into unit tests

So fixing the bad links would likely break some consumer. (I don't see the
consumer in my copy of `universe`, but if someone wrote the test, then they
must've really meant it...)

This patch adds alternative variants (`hook_links('pcp.user.actions.add')` and
`hook_links('pcp.user.actions.all')`) which should replace it.

totten committed:
(REF) hook_civicrm_links.evch.php - Move overrides

totten committed:
hook_civicrm_links.evch.php - Grandfather some more gnarly links

totten committed:
hook_civicrm_links - Add more weights:

- contribution.edit.action
- file.manage.delete
- financialItem.batch.row
- Payment.edit.action

totten commented:
I can see that this failure on the "bit" field is accurate, but the field "bit" is totally meaningless to me. Can't tell which part is wrong:

* While executing `CRM_Financial_Page_AjaxTest::testGetFinancialTransactionsList`, there's an incongruity for the list `financialItem.batch.row`.
* In the `$links`, it uses an [array-key "view"](https://github.com/civicrm/civicrm-core/blob/ad5ada56fc812d4956e7d97b824c872c98a50798/CRM/Financial/Page/BatchTransaction.php#L61). (*Aside: There are two copies of this array, per `CRM_Financial_Page_BatchTransaction::links()` and `CRM_Financial_Form_BatchTransaction::links()`. :see_no_evil:*)
* In `CRM_Core_Action`, it [copies the array-key to a property `$link['bit']`](https://github.com/civicrm/civicrm-core/blob/dce9423bdef6642986c51455ea26b50daddcb83d/CRM/Core/Action.php#L306).
* The [documentation for the hook](https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_links/) describes `bit` as "a binary number".

So somebody is wrong. Either "view" shouldn't be an array-key, or array-keys shouldn't be blindly copied to `bit`s, or the data-type isn't really "a binary number"... or "view" is a special exception that should be allowed...

totten committed:
hook_civicrm_links - Add exception for basic.CRM_Core_BAO_MessageTemplate.page

totten commented:
(___Sub-issue 2___) As per `CRM_Pledge_Form_SearchTest::testSearch()`, the links for `CRM_Pledge_Selector` [all define property `is_active`](https://github.com/civicrm/civicrm-core/blob/bd4c3e56a409050671b71832497323fd836b4f58/CRM/Pledge/Selector/Search.php#L177-L209). This property doesn't appear in the docs, and I don't think the tests flagged anywhere else. But it sounds plausible. Is it a real property (*which should be documented*) or is it someone's hallucination?

totten committed:
hook_civicrm_links - Add exception for `pledge.selector.row` (`is_active`)

totten committed:
hook_civicrm_links - Add exception for financialItem.batch.row

totten commented:
@eileenmcnaughton @colemanw - Recap: `hook_civicrm_links` dispatches in myriad ways and has random little quirks. This PR adds something to stop it from acquiring more quirks.

In the process, this PR made several of those quirks tangible. I noted a couple questionable quirks in the last two comments, but they languished for want of discussion/investigation. Which is fine. IMHO, this means we should go ahead and merge (*and just grandfather these other exceptions*).

If someone cares about this hook more than me, then they're welcome to investigate any/all anomalies. But IMHO, this PR represent progress -- it means we have  (1) a list of quirks and (2) an enforcement mechanism to prevent more.

(*Rebased on current master. Added a couple more exceptions*).

eileenmcnaughton commented:
@totten I think this is OK to merge - but the bit that adds the weights - we should target the rc for those along with the other weights fixes.

seamuslee001 commented:
@totten Maybe its just our test coverage isn't great but I would have thought this https://github.com/civicrm/civicrm-core/blob/master/CRM/Admin/Page/Extensions.php#L212 and https://github.com/civicrm/civicrm-core/blob/master/CRM/Admin/Page/Extensions.php#L259 would be triggering a test failure on Line 86 because $row['id'] here is a string (the extension key)

totten commented:
> (@eileenmcnaughton) the bit that adds the weights - we should target the rc for those along with the other weights fixes.

No problem. Cherry-picked as #27710 for 5.66-rc.

Note that one cannot remove the changes from this PR (*because the tests will fail without them*). However, since it's an exact cherry-pick and these lines aren't revised any further by the PR, they should be fairly agreeable to merges/etc.

> (@seamuslee001) Maybe its just our test coverage isn't great ...

I believe that's the explanation. Or specifically, our test-coverage leans more heavily on APIs+BAOs than on forms/screens. (*APIs+BAOs have a better story for re-use... and they're easier to test...*) I think this hook lives closer to the presentation-layer. So this patch is giving a chunk of coverage, but it's far short of total-coverage.

As the next low-hanging fruit to help that, I'd like to see is a strict-flag or dev-extension that enables all of `tests/events/*.evch.php` when using a local dev-site...

colemanw commented:
(responding to [ping](https://github.com/civicrm/civicrm-core/pull/27484#issuecomment-1744229405), I'm not just trolling but...)
From what I can tell this hook is very nearly obsolete, and will be irrelevant once we complete the work of replacing smarty/jquery tables with SearchKit displays. So I really hope this is the last of the PRs focused on this hook and we can shift our attention to that task.

> @totten As the next low-hanging fruit to help that...

How about instead of doing that, you could help me with the ongoing work of converting these tables to SearchKit & then we won't need this hook anymore. All these open PRs need review:
- #27701 
- #27700
- #27698
- https://github.com/totten/civix/pull/312

mattwire commented:
@colemanw this relates to something I came across recently with my groupadmin extension. When we convert to searchkit we lose the ability to add links via extension (in this case to the "manage groups" list) - I'm not aware of a simple way to do it via extension other than replacing the search entirely?

eileenmcnaughton commented:
@mattwire @colemanw that conversation would be good to surface higher up - perhaps a note in the docs on the alterLinks hook page?

mattwire commented:
The "Managed" hook works - see https://lab.civicrm.org/extensions/groupadmin/-/commit/dc344556381462c3e9558eaf7a28675762f4cd8f

eileenmcnaughton commented:
nice @mattwire  ... don't suppose you feel a docs update coming on?
